src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,387 @@
+package edu.kit.compiler.optimizations.common_subexpression;
+
+import com.sun.jna.Pointer;
+import edu.kit.compiler.io.DefaultWorklist;
+import edu.kit.compiler.io.Worklist;
+import edu.kit.compiler.optimizations.Optimization;
+import edu.kit.compiler.optimizations.OptimizationState;
+import edu.kit.compiler.optimizations.WorklistFiller;
+import firm.*;
+import firm.bindings.binding_irdom;
+import firm.bindings.binding_irgopt;
+import firm.bindings.binding_irnode;
+import firm.nodes.*;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+import java.util.*;
+
+@RequiredArgsConstructor
+@Data
+public class CommonSubexpressionElimination implements Optimization.Local {
+
+    // map that holds nodes that are to be exchanged with equivalent other nodes
+    private final HashMap<Node, Node> nodeValues = new HashMap<>();
+    private Worklist<Node> worklist = new DefaultWorklist<>();
+    private final Map<Pointer, Map<Pointer, Integer>> dominateMap = new HashMap<>();
+
+    @Override
+    public boolean optimize(Graph g, OptimizationState state) {
+        // compare new change map with old one
+        boolean backEdgesEnabled = BackEdges.enabled(g);
+        if (!backEdgesEnabled) {
+            BackEdges.enable(g);
+        }
+
+        boolean hadChange = false;
+        boolean changes;
+        int maxChanges = 0;
+
+        do {
+            changes = false;
+            nodeValues.clear();
+            dominateMap.clear();
+            worklist = new DefaultWorklist<>();
+
+            CSEVisitor visitor = new CSEVisitor();
+
+
+            g.walkTopological(new WorklistFiller(worklist));
+            while (!worklist.isEmpty()) {
+                Node n = worklist.dequeue();
+                n.accept(visitor);
+            }
+
+            for (var replacement : nodeValues.entrySet()) {
+                // for each replacement merge nodes together
+                Node orig = replacement.getKey();
+                Node n = replacement.getValue();
+
+                changes |= transform(g, orig, n, orig.getOpCode());
+            }
+            if (changes) {
+                hadChange = true;
+            }
+            maxChanges++;
+        } while (changes && maxChanges < 10);
+
+        if (!backEdgesEnabled) {
+            BackEdges.disable(g);
+        }
+        binding_irgopt.remove_bads(g.ptr);
+        binding_irgopt.remove_unreachable_code(g.ptr);
+        binding_irgopt.remove_bads(g.ptr);
+
+        return hadChange;
+    }
+
+    /**
+     * Transform the given node with the given associated lattice element if
+     * that element is constant.
+     */
+    private boolean transform(Graph g, Node orig, Node replacement, binding_irnode.ir_opcode opcode) {
+        Node newNode = switch (opcode) {
+            case iro_Add -> {
+                Add add = (Add) orig;
+                yield g.newAdd(orig.getBlock(), add.getLeft(), add.getRight());
+            }
```

Comment:
You have added a number of new nodes the `CSEVisitor`, don't you need to propagate these changes to `transform`?